{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "eae6af88",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<frozen importlib._bootstrap>:228: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/pandas/core/computation/expressions.py:21: UserWarning: Pandas requires version '2.8.4' or newer of 'numexpr' (version '2.8.1' currently installed).\n",
      "  from pandas.core.computation.check import NUMEXPR_INSTALLED\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import h5py\n",
    "import napari\n",
    "from scipy.sparse import issparse\n",
    "from scipy.stats import ranksums, false_discovery_control\n",
    "from tqdm import tqdm\n",
    "import pickle\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "084bcee8",
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib qt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "236ff436",
   "metadata": {},
   "outputs": [],
   "source": [
    "def hex2rgb(hexstr):\n",
    "    return tuple(int(hexstr[i:i+2], 16) for i in (0, 2, 4))\n",
    "\n",
    "\n",
    "def compute_pearson_residuals(X, theta=100.0, clip=None, \n",
    "                              copy=False, return_params=False):\n",
    "    \"\"\"from dynamo-release\"\"\"\n",
    "\n",
    "    \"\"\"Compute Pearson residuals from count data.\n",
    "\n",
    "    Pearson residuals are a measure of the deviation of observed counts from expected counts under a Poisson or negative\n",
    "    binomial model.\n",
    "\n",
    "    Args:\n",
    "        X: array_like count matrix, shape (n_cells, n_genes).\n",
    "        theta: the dispersion parameter for the negative binomial model. Must be positive.\n",
    "        clip: The maximum absolute value of the residuals. Residuals with absolute value larger than `clip` are clipped\n",
    "            to `clip`. If `None`,`clip` is set to the square root of the number of cells in `X`.\n",
    "        check_values: whether to check if `X` contains non-negative integers. If `True` and non-integer values are\n",
    "            found, a `UserWarning` is issued.\n",
    "        copy: whether to make a copy of `X`.\n",
    "\n",
    "    Returns:\n",
    "        The Pearson residuals.\n",
    "    \"\"\"\n",
    "    X = X.copy() if copy else X\n",
    "\n",
    "    # check theta\n",
    "    if theta <= 0:\n",
    "        # TODO: would \"underdispersion\" with negative theta make sense?\n",
    "        # then only theta=0 were undefined..\n",
    "        raise ValueError(\"Pearson residuals require theta > 0\")\n",
    "    # prepare clipping\n",
    "    if clip is None:\n",
    "        n = X.shape[0]\n",
    "        clip = np.sqrt(n)\n",
    "    if clip < 0:\n",
    "        raise ValueError(\"Pearson residuals require `clip>=0` or `clip=None`.\")\n",
    "\n",
    "    if issparse(X):\n",
    "        sums_genes = np.sum(X, axis=0)\n",
    "        sums_cells = np.sum(X, axis=1)\n",
    "        sum_total = np.sum(sums_genes).squeeze()\n",
    "    else:\n",
    "        sums_genes = np.sum(X, axis=0, keepdims=True)\n",
    "        sums_cells = np.sum(X, axis=1, keepdims=True)\n",
    "        sum_total = np.sum(sums_genes)\n",
    "\n",
    "    mu = np.array(sums_cells @ sums_genes / sum_total)\n",
    "    diff = np.array(X - mu)\n",
    "    residuals = diff / np.sqrt(mu + mu**2 / theta)\n",
    "\n",
    "    # clip\n",
    "    residuals = np.clip(residuals, a_min=-clip, a_max=clip)\n",
    "    \n",
    "    if return_params:\n",
    "        sigma = np.sqrt(mu + mu**2 / theta)\n",
    "        return residuals, mu, sigma\n",
    "    else:\n",
    "        return residuals\n",
    "\n",
    "    \n",
    "def logxp1(counts):\n",
    "    \"\"\"return normalized log(x+1) transform of cellxgene matrix of raw counts\"\"\"\n",
    "    return np.log(counts / np.expand_dims(np.sum(counts, axis=1), axis=1) + 1)\n",
    "    \n",
    "    \n",
    "def bin_aps(aps, bins):\n",
    "    _counts, bins = np.histogram(aps, bins)\n",
    "    bins = bins[1:]\n",
    "    binned_aps = np.zeros_like(aps)\n",
    "    for i in range(len(binned_aps)):\n",
    "        binned_aps[i] = get_ap_bin(aps[i], bins)\n",
    "    \n",
    "    return binned_aps\n",
    "    \n",
    "\n",
    "def get_ap_bin(this_ap, bins):\n",
    "    this_bin = np.where(np.abs(this_ap - bins) == np.nanmin(np.abs(this_ap - bins)))[0][0]\n",
    "\n",
    "    return this_bin\n",
    "\n",
    "\n",
    "def bin_exp(exp, binned_ys, y_bins, n_bootstraps=300):\n",
    "    binned_exp = np.zeros_like(y_bins)\n",
    "    std_exp = np.zeros_like(y_bins)\n",
    "    for i in range(len(y_bins)):\n",
    "        these_expression_levels = exp[y_bins[binned_ys.astype('int')] == y_bins[i]]\n",
    "        binned_exp[i] = np.nanmean(these_expression_levels)\n",
    "        bootstrapped_means = np.zeros(n_bootstraps)\n",
    "        for n in range(n_bootstraps):\n",
    "            these_expression_levels = np.random.choice(these_expression_levels, len(these_expression_levels))\n",
    "            bootstrapped_means[n] = np.mean(these_expression_levels)\n",
    "        std_exp[i] = np.nanstd(bootstrapped_means)\n",
    "    return binned_exp, std_exp\n",
    "\n",
    "\n",
    "def get_filtered_index(index, gene_list, filtered_gene_list):\n",
    "    this_gene = gene_list[index]\n",
    "    filtered_index = np.where(np.array([g == this_gene for g in filtered_gene_list]))[0][0]\n",
    "    \n",
    "    return filtered_index\n",
    "    \n",
    "    \n",
    "def get_binned_line_dist(index, residuals, binned_ys, y_bins, gene_list, filtered_gene_list):\n",
    "    filtered_index = get_filtered_index(index, gene_list, filtered_gene_list)\n",
    "    expression = residuals[:, filtered_index]\n",
    "    binned_expression, _ = bin_exp(expression, binned_ys, y_bins)\n",
    "    \n",
    "    return binned_expression\n",
    "    \n",
    "\n",
    "def get_line_dist_uncertainty(index, residuals, binned_ys, y_bins, gene_list, filtered_gene_list, n_bootstraps=100):\n",
    "    line_dist_arr = np.zeros((n_bootstraps, len(y_bins)))\n",
    "    for i in range(n_bootstraps):\n",
    "        scrambled_ys = np.random.choice(binned_ys, len(binned_ys))\n",
    "        line_dist_arr[i] = get_binned_line_dist(index, residuals, scrambled_ys, y_bins, gene_list, filtered_gene_list)\n",
    "\n",
    "    return np.nanmean(line_dist_arr, axis=0), np.nanstd(line_dist_arr, axis=0)\n",
    "        \n",
    "\n",
    "def compute_anterior_middle_difference(gene_id, X, \n",
    "                                  binned_ys, y_bins, \n",
    "                                  gene_list, filtered_gene_list, \n",
    "                                  n_bootstraps=300, a_start=-113.0, \n",
    "                                  a_end=-63, m_start=0, m_end=50):\n",
    "    \n",
    "    exp = X[:, get_filtered_index(gene_id, gene_list, \n",
    "                                          filtered_gene_list)]\n",
    "    binned_exp, std_exp = bin_exp(exp, binned_ys, y_bins, \n",
    "                                  n_bootstraps=n_bootstraps)\n",
    "    mean_scramble, std_scramble = get_line_dist_uncertainty(\n",
    "                gene_id, X, binned_ys, y_bins, gene_list, \n",
    "                filtered_gene_list, n_bootstraps=n_bootstraps)\n",
    "    \n",
    "    anterior_ids = (y_bins >= a_start) & (y_bins < a_end)\n",
    "    middle_ids = (y_bins >= m_start) & (y_bins < m_end)\n",
    "\n",
    "    anterior_signal = np.mean(binned_exp[anterior_ids])\n",
    "    middle_signal = np.mean(binned_exp[middle_ids])\n",
    "    \n",
    "    difference = anterior_signal - middle_signal\n",
    "    sigma_anterior = np.sqrt(np.mean(std_exp[anterior_ids] ** 2) / np.sum(anterior_ids))\n",
    "    sigma_middle = np.sqrt(np.mean(std_exp[middle_ids] ** 2))\n",
    "    sigma_difference = np.sqrt((sigma_anterior ** 2 + sigma_middle **2) / np.sum(middle_ids))\n",
    "    \n",
    "\n",
    "    anterior_scramble = np.mean(mean_scramble[anterior_ids])\n",
    "    middle_scramble = np.mean(mean_scramble[middle_ids])\n",
    "    \n",
    "    scramble_difference = anterior_scramble - middle_scramble\n",
    "    sigma_scramble_anterior = np.sqrt(np.mean(std_scramble[anterior_ids] ** 2) / np.sum(anterior_ids))\n",
    "    sigma_scramble_middle = np.sqrt(np.mean(std_scramble[middle_ids] ** 2) / np.sum(middle_ids))\n",
    "    sigma_scramble_difference = np.sqrt((sigma_scramble_anterior ** 2 + sigma_scramble_middle **2))\n",
    "    \n",
    "    \n",
    "    \n",
    "    return difference, sigma_difference, scramble_difference, sigma_scramble_difference\n",
    "\n",
    "        \n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3a7b32bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<KeysViewHDF5 ['X', 'layers', 'obs', 'obsm', 'uns', 'var']>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"load the anndata file\"\"\"\n",
    "file_path = r'/media/brandon/Data2/Brandon/fly_immune/Flysta3d/L3_b_count_normal_stereoseq.h5ad'\n",
    "f = h5py.File(file_path, 'r')\n",
    "f.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c07a6f7b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3195445/1622129586.py:2: DeprecationWarning: `product` is deprecated as of NumPy 1.25.0, and will be removed in NumPy 2.0. Please use `prod` instead.\n",
      "  raw_reads =  np.array(f['layers'].get('raw_counts'))\n",
      "/tmp/ipykernel_3195445/625980817.py:51: RuntimeWarning: invalid value encountered in divide\n",
      "  residuals = diff / np.sqrt(mu + mu**2 / theta)\n"
     ]
    }
   ],
   "source": [
    "\"\"\"extract the raw reads and processes\"\"\"\n",
    "raw_reads =  np.array(f['layers'].get('raw_counts'))\n",
    "\n",
    "# filter reads to 5% detection\n",
    "detection_percent = np.sum(raw_reads > 0, axis=0) / len(raw_reads)\n",
    "filter_sel = detection_percent > 0.05\n",
    "filtered_reads = raw_reads[:, filter_sel]\n",
    "\n",
    "# convert to pearson residuals\n",
    "residuals = compute_pearson_residuals(filtered_reads, theta=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1e48fa1d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3195445/2013191573.py:2: DeprecationWarning: `product` is deprecated as of NumPy 1.25.0, and will be removed in NumPy 2.0. Please use `prod` instead.\n",
      "  a = np.array(f['obs'].get('annotation'))\n"
     ]
    }
   ],
   "source": [
    "\"\"\"get selection of just fat body cells\"\"\"\n",
    "a = np.array(f['obs'].get('annotation'))\n",
    "sel = (a == 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c4e6524b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fat body cells range from y=-178.21475 to y=184.0076\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3195445/1694782897.py:2: DeprecationWarning: `product` is deprecated as of NumPy 1.25.0, and will be removed in NumPy 2.0. Please use `prod` instead.\n",
      "  spatial = np.array(f['obsm'].get('spatial'))\n"
     ]
    }
   ],
   "source": [
    "\"\"\"get the y coordinates of fat body cells\"\"\"\n",
    "spatial = np.array(f['obsm'].get('spatial'))\n",
    "spatial = spatial[sel, :]\n",
    "\n",
    "ys = spatial[:, 1]\n",
    "print(f'fat body cells range from y={np.min(ys)} to y={np.max(ys)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "bd670c12",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"load the gene names and convert to list\"\"\"\n",
    "gene_ids = f['var'].get('geneID')\n",
    "gene_indices = np.arange(len(gene_ids))\n",
    "\n",
    "\"\"\"convert gene_ids to more convenient lists\"\"\"\n",
    "gene_list = [gene.decode() for gene in gene_ids]\n",
    "filtered_gene_list = [gene.decode() for gene in gene_ids[filter_sel]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0642875a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# parameters for ap bins\n",
    "n_y_bins = 30\n",
    "y_bins = np.linspace(np.min(ys), np.max(ys), n_y_bins)\n",
    "binned_ys = bin_aps(ys, y_bins)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "093ccf17",
   "metadata": {},
   "source": [
    "# Identify genes with positive or negative anterior bias"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "8283baad",
   "metadata": {},
   "outputs": [],
   "source": [
    "run = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "85d55107",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"y coordinates that define the start and end of anterior and middle regions\"\"\"\n",
    "a_start = -113\n",
    "a_end = -113 + 50\n",
    "\n",
    "m_start = 0\n",
    "m_end = 50\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "aaf21d48",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'filtered_gene_list' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [15]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0m all_differences \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mzeros(\u001b[38;5;28mlen\u001b[39m(\u001b[43mfiltered_gene_list\u001b[49m))\n\u001b[1;32m      2\u001b[0m all_sigma_differences \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mzeros(\u001b[38;5;28mlen\u001b[39m(filtered_gene_list))\n\u001b[1;32m      3\u001b[0m all_scramble_differences \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mzeros(\u001b[38;5;28mlen\u001b[39m(filtered_gene_list))\n",
      "\u001b[0;31mNameError\u001b[0m: name 'filtered_gene_list' is not defined"
     ]
    }
   ],
   "source": [
    "\"\"\"run the computation of residual differences. Takes ~2.5 hrs\"\"\"\n",
    "if run:\n",
    "    all_differences = np.zeros(len(filtered_gene_list))\n",
    "    all_sigma_differences = np.zeros(len(filtered_gene_list))\n",
    "    all_scramble_differences = np.zeros(len(filtered_gene_list))\n",
    "    all_sigma_scramble_differences = np.zeros(len(filtered_gene_list))\n",
    "\n",
    "\n",
    "    for i in tqdm(range(len(filtered_gene_list))):\n",
    "        this_id = np.where(np.array([g == filtered_gene_list[i] for g in gene_list]))[0][0]\n",
    "        all_differences[i], \n",
    "        all_sigma_differences[i], \n",
    "        all_scramble_differences[i], \n",
    "        all_sigma_scramble_differences[i] = compute_anterior_middle_difference(this_id, residuals[sel], \n",
    "                                      binned_ys, y_bins, \n",
    "                                      gene_list, filtered_gene_list, \n",
    "                                      n_bootstraps=10, a_start=a_start, \n",
    "                                      a_end=a_end, m_start=m_start, m_end=m_end)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "f192bb57",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"load the results\"\"\"\n",
    "with open(r'/media/brandon/Data2/Brandon/fly_immune/Flysta3d/anterior-middle-differences.pkl', 'rb') as file:\n",
    "    all_differences, all_sigma_differences, all_scramble_differences, all_sigma_scramble_differences  =  pickle.load(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "0a978166",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"filter the results\"\"\"\n",
    "# filter by error bars\n",
    "high_expressing = all_differences - all_sigma_differences > (all_scramble_differences + all_sigma_scramble_differences)\n",
    "low_expressing = all_differences + all_sigma_differences < (all_scramble_differences - all_sigma_scramble_differences)\n",
    "\n",
    "# filter by effect size\n",
    "difference_thresh = 2\n",
    "high_expressing = high_expressing * (all_differences > difference_thresh)\n",
    "low_expressing = low_expressing * (all_differences < -1 * difference_thresh)\n",
    "\n",
    "# extract the gene names that survived the filter\n",
    "high_expressing_genes = [filtered_gene_list[i] for i in range(len(filtered_gene_list)) if high_expressing[i]]\n",
    "low_expressing_genes = [filtered_gene_list[i] for i in range(len(filtered_gene_list)) if low_expressing[i]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "78dab810",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1902, CG14302\n",
      "2329, CG15369\n",
      "4737, CG42587\n",
      "5979, CG6870\n",
      "6276, CG7953\n",
      "7808, Fbp1\n",
      "8682, LysS\n",
      "10822, TotA\n",
      "12109, deltaTry\n",
      "12787, lncRNA:CR33938\n",
      "12797, lncRNA:CR34335\n",
      "12802, lncRNA:CR40469\n",
      "15171, regucalcin\n"
     ]
    }
   ],
   "source": [
    "\"\"\"print the gene names\"\"\"\n",
    "for g in high_expressing_genes:\n",
    "    this_id = np.where(np.array([gg == g for gg in gene_list]))[0][0]\n",
    "\n",
    "    print(f'{gene_indices[this_id]}, {g}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "3958eac2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1104, CG12116\n",
      "1211, CG12522\n",
      "1254, CG12699\n",
      "1628, CG13641\n",
      "2548, CG16713\n",
      "2603, CG16926\n",
      "2729, CG17376\n",
      "3296, CG30430\n",
      "3409, CG31226\n",
      "3440, CG31313\n",
      "3648, CG31988\n",
      "3672, CG32073\n",
      "4216, CG34166\n",
      "4218, CG34168\n",
      "5341, CG46059\n",
      "5418, CG4716\n",
      "5700, CG5773\n",
      "6918, CR43186\n",
      "8671, Lsp1beta\n",
      "8672, Lsp1gamma\n",
      "8673, Lsp2\n",
      "8896, Mst84Db\n",
      "8898, Mst84Dd\n",
      "8900, Mst87F\n",
      "9229, ORY\n",
      "10150, S-Lap1\n",
      "10153, S-Lap4\n",
      "11345, asRNA:CR11538\n",
      "14921, ocn\n",
      "15101, pre-rRNA:CR45856\n"
     ]
    }
   ],
   "source": [
    "\"\"\"print the gene names\"\"\"\n",
    "for g in low_expressing_genes:\n",
    "    this_id = np.where(np.array([gg == g for gg in gene_list]))[0][0]\n",
    "\n",
    "    print(f'{gene_indices[this_id]}, {g}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "358692cd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/core/fromnumeric.py:3474: RuntimeWarning: Mean of empty slice.\n",
      "  return _methods._mean(a, axis=axis, dtype=dtype,\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/core/_methods.py:189: RuntimeWarning: invalid value encountered in true_divide\n",
      "  ret = ret.dtype.type(ret / rcount)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:114: RuntimeWarning: Mean of empty slice\n",
      "  return np.nanmean(line_dist_arr, axis=0), np.nanstd(line_dist_arr, axis=0)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Text(0, 0.5, 'DptA')"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#this_id = 10822    # totA\n",
    "#this_id = 15171    # regulacin\n",
    "#this_id = 8673 \n",
    "this_id = 7604      # dptA\n",
    "expression = residuals[sel, get_filtered_index(this_id, gene_list, filtered_gene_list)]\n",
    "binned_exp, std_exp = bin_exp(expression, binned_ys, y_bins)\n",
    "mean_scramble, std_scramble = get_line_dist_uncertainty(this_id, residuals[sel], binned_ys, y_bins, gene_list, filtered_gene_list, n_bootstraps=10)\n",
    "m = mean_scramble\n",
    "s = std_scramble\n",
    "l = m - s\n",
    "u = m + s\n",
    "\n",
    "plt.figure()\n",
    "plt.fill_between(y_bins, l, u, color='k', alpha=0.3)\n",
    "plt.errorbar(y_bins, binned_exp, std_exp, linewidth=4, color='c')\n",
    "#plt.errorbar(y_bins, binned_dptA, bootstrapped_uncertainty)\n",
    "\n",
    "plt.xlabel('ap')\n",
    "plt.ylabel(f'{gene_list[this_id]}')\n",
    "#plt.xlim([-113, np.max(ys)])\n",
    "#plt.ylim([-1.0, 2.5])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "370e345d",
   "metadata": {},
   "source": [
    "## Make plots of all up and down regulated gene line dists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "43473dcf",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "0it [00:00, ?it/s]/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "13it [00:00, 213.75it/s]\n"
     ]
    }
   ],
   "source": [
    "\"\"\"up regulated\"\"\"\n",
    "plt.figure()\n",
    "\n",
    "for i, g in tqdm(enumerate(high_expressing_genes)):\n",
    "    this_id = np.where(np.array([gg == g for gg in gene_list]))[0][0]\n",
    "    expression = residuals[sel, get_filtered_index(this_id, gene_list, filtered_gene_list)]\n",
    "    binned_exp, _ = bin_exp(expression, binned_ys, y_bins, n_bootstraps=0)\n",
    "    plt.plot(y_bins, binned_exp - np.nanmin(binned_exp), linewidth=4)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "b9d983fd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "0it [00:00, ?it/s]/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24it [00:00, 231.76it/s]/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_2504142/1669947554.py:84: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "30it [00:00, 235.71it/s]\n"
     ]
    }
   ],
   "source": [
    "\"\"\"down regulated\"\"\"\n",
    "plt.figure()\n",
    "\n",
    "for i, g in tqdm(enumerate(low_expressing_genes)):\n",
    "    this_id = np.where(np.array([gg == g for gg in gene_list]))[0][0]\n",
    "    expression = residuals[sel, get_filtered_index(this_id, gene_list, filtered_gene_list)]\n",
    "    binned_exp, _ = bin_exp(expression, binned_ys, y_bins, n_bootstraps=0)\n",
    "    plt.plot(y_bins, binned_exp - np.nanmin(binned_exp), linewidth=4)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "314f4fca",
   "metadata": {},
   "source": [
    "## Standard approach\n",
    "Normalize by read depth per cell and logp1 transform"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "4a9be148",
   "metadata": {},
   "outputs": [],
   "source": [
    "run = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b388e2ae",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3195445/4111918620.py:2: DeprecationWarning: `product` is deprecated as of NumPy 1.25.0, and will be removed in NumPy 2.0. Please use `prod` instead.\n",
      "  raw_reads =  np.array(f['layers'].get('raw_counts'))\n"
     ]
    }
   ],
   "source": [
    "\"\"\"construct array of normalized expression levels filtered for detection and fat body markers\"\"\"\n",
    "raw_reads =  np.array(f['layers'].get('raw_counts'))\n",
    "\n",
    "norm_reads = raw_reads / np.expand_dims(np.sum(raw_reads, axis=1), axis=1)\n",
    "\n",
    "X = logxp1(raw_reads)\n",
    "\n",
    "# filter reads to 5% detection\n",
    "detection_percent = np.sum(raw_reads > 0, axis=0) / len(raw_reads)\n",
    "filter_sel = detection_percent > 0.05\n",
    "filtered_reads = raw_reads[:, filter_sel]\n",
    "\n",
    "X_filtered = X[:, filter_sel]\n",
    "X_fb = X_filtered[sel]\n",
    "norm_reads_filtered = norm_reads[:, filter_sel]\n",
    "norm_reads_fb = norm_reads_filtered[sel]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c51eda12",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0, 0.5, 'var count')"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check variance-mean relationship.\n",
    "plt.figure()\n",
    "plt.plot(np.mean(X_fb, axis=0), np.var(X_fb, axis=0), 'ko', markerfacecolor='c', markersize=4, alpha=0.1)\n",
    "x = np.logspace(-5, -1)\n",
    "plt.plot(x, x, 'm--', linewidth=2)\n",
    "plt.xscale('log')\n",
    "plt.yscale('log')\n",
    "plt.xlabel('mean count')\n",
    "plt.ylabel('var count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1cf4c5c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1079\n",
      "1470\n"
     ]
    }
   ],
   "source": [
    "\"\"\"y coordinates that define the start and end of anterior and middle regions\"\"\"\n",
    "a_start = -113\n",
    "a_end = a_start + 70\n",
    "\n",
    "m_start = 0\n",
    "m_end = m_start + 70\n",
    "\n",
    "anterior_ids = (ys >= a_start) & (ys < a_end)\n",
    "middle_ids = (ys >= m_start) & (ys < m_end)\n",
    "\n",
    "print(np.sum(anterior_ids))\n",
    "print(np.sum(middle_ids))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0eee0520",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████████| 6505/6505 [00:16<00:00, 401.84it/s]\n"
     ]
    }
   ],
   "source": [
    "\"\"\"run the computation of differential expression analysis. \"\"\"\n",
    "n_bootstraps = 10\n",
    "if run:\n",
    "    all_logfc = np.zeros(len(filtered_gene_list))\n",
    "    all_pvalues = np.zeros(len(filtered_gene_list))\n",
    "    all_means = np.zeros(len(filtered_gene_list))\n",
    "    all_uncertainties = np.zeros(len(filtered_gene_list))\n",
    "    all_scrambled_means = np.zeros(len(filtered_gene_list))\n",
    "    all_scrambled_uncertainties = np.zeros(len(filtered_gene_list))\n",
    "    all_middle_means = np.zeros(len(filtered_gene_list))\n",
    "    all_middle_uncertainties = np.zeros(len(filtered_gene_list))\n",
    "    for i in tqdm(range(len(filtered_gene_list))):\n",
    "        anterior_expression = X_fb[anterior_ids, i]\n",
    "        middle_expression = X_fb[middle_ids, i]\n",
    "        _, all_pvalues[i] = ranksums(anterior_expression, middle_expression)\n",
    "        #all_logfc[i] = np.log(np.mean(anterior_expression) / np.mean(middle_expression))\n",
    "        \n",
    "        anterior_norm = norm_reads_fb[anterior_ids, i]\n",
    "        middle_norm = norm_reads_fb[middle_ids, i]\n",
    "        all_logfc[i] = np.log2(np.mean(anterior_norm) / np.mean(middle_norm))\n",
    "        \n",
    "        # real mean\n",
    "        all_means[i] = np.mean(anterior_norm)\n",
    "\n",
    "        # get real uncertainty\n",
    "        bootstrapped_mean = np.zeros(n_bootstraps)\n",
    "        for n in range(n_bootstraps):\n",
    "            bootstrapped_mean[n] = np.mean(np.random.choice(anterior_norm, len(anterior_ids)))\n",
    "        \n",
    "        all_uncertainties[i] = np.std(bootstrapped_mean)\n",
    "\n",
    "        # real middle mean\n",
    "        all_middle_means[i] = np.mean(middle_norm)\n",
    "\n",
    "        # get real middle uncertainty\n",
    "        bootstrapped_mean = np.zeros(n_bootstraps)\n",
    "        for n in range(n_bootstraps):\n",
    "            bootstrapped_mean[n] = np.mean(np.random.choice(middle_norm, len(anterior_ids)))\n",
    "        \n",
    "        all_middle_uncertainties[i] = np.std(bootstrapped_mean)\n",
    "        \n",
    "        # get scrambled mean and uncertainty\n",
    "        bootstrapped_mean = np.zeros(n_bootstraps)\n",
    "        for n in range(n_bootstraps):\n",
    "            bootstrapped_mean[n] = np.mean(np.random.choice(norm_reads_fb[:, i], len(anterior_ids)))\n",
    "        \n",
    "        all_scrambled_means[i] = np.mean(bootstrapped_mean)\n",
    "        all_scrambled_uncertainties[i] = np.std(bootstrapped_mean)\n",
    "        \n",
    "            \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "69232b50",
   "metadata": {},
   "outputs": [],
   "source": [
    "lfc_thresh = 1#0.5#np.log2(2)\n",
    "p_thresh = 0.05 / len(filtered_gene_list)\n",
    "all_corrected_pvalues = all_pvalues #false_discovery_control(all_pvalues)\n",
    "up_sel = (all_corrected_pvalues < p_thresh) & (all_logfc > lfc_thresh)\n",
    "down_sel = (all_corrected_pvalues < p_thresh) & (all_logfc < -lfc_thresh)\n",
    "diffex_sel = (all_corrected_pvalues < p_thresh) & (np.abs(all_logfc) > lfc_thresh)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "b41a53fe",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3195445/4144405191.py:4: RuntimeWarning: divide by zero encountered in log10\n",
      "  plt.plot(plot_x, -np.log10(plot_y), 'ko', markersize=6, alpha=0.5)\n"
     ]
    }
   ],
   "source": [
    "plt.figure()\n",
    "plot_x = all_logfc[diffex_sel]\n",
    "plot_y = all_pvalues[diffex_sel]\n",
    "plt.plot(plot_x, -np.log10(plot_y), 'ko', markersize=6, alpha=0.5)\n",
    "plt.yscale('linear')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "ddf4e134",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"assemble results into a dataframe\"\"\"\n",
    "df = pd.DataFrame()\n",
    "df['gene'] = filtered_gene_list\n",
    "df['p_val'] = all_pvalues\n",
    "df['log2fc'] = all_logfc\n",
    "df['anterior_mean'] = all_means\n",
    "df['anterior_uncertainty'] = all_uncertainties\n",
    "df['middle_mean'] = all_middle_means\n",
    "df['middle_uncertainty'] = all_middle_uncertainties\n",
    "df['scrambled_mean'] = all_scrambled_means\n",
    "df['scrambled_uncertainty'] = all_scrambled_uncertainties"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "99c73498",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"filter on p value, fold change, and expression relative to a uniform null\"\"\"\n",
    "anterior_df = df[(df.p_val < p_thresh) & (df.log2fc > lfc_thresh) & (df.anterior_mean - df.anterior_uncertainty > 1.5 * (df.scrambled_mean + df.scrambled_uncertainty))]\n",
    "middle_df = df[(df.p_val < p_thresh) & (df.log2fc < -lfc_thresh) & (df.middle_mean - df.middle_uncertainty > 1.5 * (df.scrambled_mean + df.scrambled_uncertainty))]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "eecd9e9d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>r4gal4_v60100_0_15m_1</th>\n",
       "      <th>r4gal4_v60100_0_15m_2</th>\n",
       "      <th>r4gal4_v60100_0_15m_3</th>\n",
       "      <th>r4gal4_svpi_0_15m_1</th>\n",
       "      <th>r4gal4_svpi_0_15m_2</th>\n",
       "      <th>r4gal4_svpi_0_15m_3</th>\n",
       "      <th>r4gal4_v60100_0_7m_1</th>\n",
       "      <th>r4gal4_v60100_0_7m_2</th>\n",
       "      <th>r4gal4_v60100_0_7m_3</th>\n",
       "      <th>r4gal4_svpi_0_7m_1</th>\n",
       "      <th>r4gal4_svpi_0_7m_2</th>\n",
       "      <th>r4gal4_svpi_0_7m_3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Alas</td>\n",
       "      <td>288.58000</td>\n",
       "      <td>163.1200</td>\n",
       "      <td>186.1700</td>\n",
       "      <td>226.95000</td>\n",
       "      <td>205.9600</td>\n",
       "      <td>177.2200</td>\n",
       "      <td>320.1900</td>\n",
       "      <td>199.9600</td>\n",
       "      <td>319.3900</td>\n",
       "      <td>193.55000</td>\n",
       "      <td>222.2200</td>\n",
       "      <td>172.1100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>alc</td>\n",
       "      <td>130.61000</td>\n",
       "      <td>146.0300</td>\n",
       "      <td>155.2300</td>\n",
       "      <td>163.78000</td>\n",
       "      <td>163.0600</td>\n",
       "      <td>161.8100</td>\n",
       "      <td>185.6400</td>\n",
       "      <td>155.1600</td>\n",
       "      <td>141.8000</td>\n",
       "      <td>218.14000</td>\n",
       "      <td>202.9800</td>\n",
       "      <td>215.9600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AlCR2</td>\n",
       "      <td>15.39500</td>\n",
       "      <td>1.3285</td>\n",
       "      <td>6.2577</td>\n",
       "      <td>0.96298</td>\n",
       "      <td>9.4032</td>\n",
       "      <td>10.1400</td>\n",
       "      <td>10.6350</td>\n",
       "      <td>7.7955</td>\n",
       "      <td>8.2266</td>\n",
       "      <td>2.83440</td>\n",
       "      <td>2.3485</td>\n",
       "      <td>2.5456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ald</td>\n",
       "      <td>48.98900</td>\n",
       "      <td>34.5790</td>\n",
       "      <td>19.3240</td>\n",
       "      <td>36.49200</td>\n",
       "      <td>41.1770</td>\n",
       "      <td>45.5720</td>\n",
       "      <td>35.2910</td>\n",
       "      <td>44.9340</td>\n",
       "      <td>33.9270</td>\n",
       "      <td>38.86500</td>\n",
       "      <td>26.7900</td>\n",
       "      <td>22.7610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Ald</td>\n",
       "      <td>48.98900</td>\n",
       "      <td>34.5790</td>\n",
       "      <td>19.3240</td>\n",
       "      <td>36.49200</td>\n",
       "      <td>41.1770</td>\n",
       "      <td>45.5720</td>\n",
       "      <td>2595.8000</td>\n",
       "      <td>1966.5000</td>\n",
       "      <td>2609.0000</td>\n",
       "      <td>1851.70000</td>\n",
       "      <td>1638.1000</td>\n",
       "      <td>1829.1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9522</th>\n",
       "      <td>zuc</td>\n",
       "      <td>0.82583</td>\n",
       "      <td>1.1566</td>\n",
       "      <td>4.0080</td>\n",
       "      <td>2.95120</td>\n",
       "      <td>5.0812</td>\n",
       "      <td>4.9883</td>\n",
       "      <td>3.2505</td>\n",
       "      <td>3.8929</td>\n",
       "      <td>7.1624</td>\n",
       "      <td>0.87707</td>\n",
       "      <td>7.8542</td>\n",
       "      <td>4.2576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9523</th>\n",
       "      <td>Zw</td>\n",
       "      <td>85.34800</td>\n",
       "      <td>111.5100</td>\n",
       "      <td>145.1500</td>\n",
       "      <td>67.09700</td>\n",
       "      <td>88.9270</td>\n",
       "      <td>37.2600</td>\n",
       "      <td>200.5100</td>\n",
       "      <td>139.5100</td>\n",
       "      <td>150.0000</td>\n",
       "      <td>154.82000</td>\n",
       "      <td>140.5700</td>\n",
       "      <td>186.4200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9524</th>\n",
       "      <td>zwilch</td>\n",
       "      <td>35.36500</td>\n",
       "      <td>30.2110</td>\n",
       "      <td>36.8080</td>\n",
       "      <td>18.72800</td>\n",
       "      <td>29.7150</td>\n",
       "      <td>21.3560</td>\n",
       "      <td>24.6340</td>\n",
       "      <td>35.1830</td>\n",
       "      <td>24.6810</td>\n",
       "      <td>28.38700</td>\n",
       "      <td>26.5840</td>\n",
       "      <td>19.3860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9525</th>\n",
       "      <td>zye</td>\n",
       "      <td>339.24000</td>\n",
       "      <td>365.5200</td>\n",
       "      <td>306.0800</td>\n",
       "      <td>259.56000</td>\n",
       "      <td>183.2200</td>\n",
       "      <td>110.6100</td>\n",
       "      <td>247.6900</td>\n",
       "      <td>231.2900</td>\n",
       "      <td>230.1000</td>\n",
       "      <td>146.30000</td>\n",
       "      <td>148.5000</td>\n",
       "      <td>172.6400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9526</th>\n",
       "      <td>Zyx102EF</td>\n",
       "      <td>206.14000</td>\n",
       "      <td>155.1100</td>\n",
       "      <td>154.8600</td>\n",
       "      <td>199.29000</td>\n",
       "      <td>230.4500</td>\n",
       "      <td>218.0000</td>\n",
       "      <td>166.8100</td>\n",
       "      <td>199.0700</td>\n",
       "      <td>263.8700</td>\n",
       "      <td>256.11000</td>\n",
       "      <td>210.9300</td>\n",
       "      <td>153.0200</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9527 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Gene  r4gal4_v60100_0_15m_1  r4gal4_v60100_0_15m_2  \\\n",
       "0         Alas              288.58000               163.1200   \n",
       "1          alc              130.61000               146.0300   \n",
       "2        AlCR2               15.39500                 1.3285   \n",
       "3          ald               48.98900                34.5790   \n",
       "4          Ald               48.98900                34.5790   \n",
       "...        ...                    ...                    ...   \n",
       "9522       zuc                0.82583                 1.1566   \n",
       "9523        Zw               85.34800               111.5100   \n",
       "9524    zwilch               35.36500                30.2110   \n",
       "9525       zye              339.24000               365.5200   \n",
       "9526  Zyx102EF              206.14000               155.1100   \n",
       "\n",
       "      r4gal4_v60100_0_15m_3  r4gal4_svpi_0_15m_1  r4gal4_svpi_0_15m_2  \\\n",
       "0                  186.1700            226.95000             205.9600   \n",
       "1                  155.2300            163.78000             163.0600   \n",
       "2                    6.2577              0.96298               9.4032   \n",
       "3                   19.3240             36.49200              41.1770   \n",
       "4                   19.3240             36.49200              41.1770   \n",
       "...                     ...                  ...                  ...   \n",
       "9522                 4.0080              2.95120               5.0812   \n",
       "9523               145.1500             67.09700              88.9270   \n",
       "9524                36.8080             18.72800              29.7150   \n",
       "9525               306.0800            259.56000             183.2200   \n",
       "9526               154.8600            199.29000             230.4500   \n",
       "\n",
       "      r4gal4_svpi_0_15m_3  r4gal4_v60100_0_7m_1  r4gal4_v60100_0_7m_2  \\\n",
       "0                177.2200              320.1900              199.9600   \n",
       "1                161.8100              185.6400              155.1600   \n",
       "2                 10.1400               10.6350                7.7955   \n",
       "3                 45.5720               35.2910               44.9340   \n",
       "4                 45.5720             2595.8000             1966.5000   \n",
       "...                   ...                   ...                   ...   \n",
       "9522               4.9883                3.2505                3.8929   \n",
       "9523              37.2600              200.5100              139.5100   \n",
       "9524              21.3560               24.6340               35.1830   \n",
       "9525             110.6100              247.6900              231.2900   \n",
       "9526             218.0000              166.8100              199.0700   \n",
       "\n",
       "      r4gal4_v60100_0_7m_3  r4gal4_svpi_0_7m_1  r4gal4_svpi_0_7m_2  \\\n",
       "0                 319.3900           193.55000            222.2200   \n",
       "1                 141.8000           218.14000            202.9800   \n",
       "2                   8.2266             2.83440              2.3485   \n",
       "3                  33.9270            38.86500             26.7900   \n",
       "4                2609.0000          1851.70000           1638.1000   \n",
       "...                    ...                 ...                 ...   \n",
       "9522                7.1624             0.87707              7.8542   \n",
       "9523              150.0000           154.82000            140.5700   \n",
       "9524               24.6810            28.38700             26.5840   \n",
       "9525              230.1000           146.30000            148.5000   \n",
       "9526              263.8700           256.11000            210.9300   \n",
       "\n",
       "      r4gal4_svpi_0_7m_3  \n",
       "0               172.1100  \n",
       "1               215.9600  \n",
       "2                 2.5456  \n",
       "3                22.7610  \n",
       "4              1829.1000  \n",
       "...                  ...  \n",
       "9522              4.2576  \n",
       "9523            186.4200  \n",
       "9524             19.3860  \n",
       "9525            172.6400  \n",
       "9526            153.0200  \n",
       "\n",
       "[9527 rows x 13 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"futher filter by cross referencing a list of genes expressed in the larval fat body by bulk RNA seq\"\"\"\n",
    "bulk_df = pd.read_excel(r'/home/brandon/Downloads/GSE95800_GEO_upload_processed_data.xlsx')\n",
    "bulk_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "d4277704",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"extract list of genes in the bulk data set. replace known name discrepencies\"\"\"\n",
    "bulk_genes = bulk_df.Gene.to_list()\n",
    "bulk_genes = list(map(lambda x: x.replace('Dpt', 'DptA'), bulk_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "a1c2c561",
   "metadata": {},
   "outputs": [],
   "source": [
    "anterior_df = anterior_df[[g in bulk_genes for g in anterior_df.gene.to_list()]]\n",
    "middle_df = middle_df[[g in bulk_genes for g in middle_df.gene.to_list()]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "8aa7c833",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>p_val</th>\n",
       "      <th>log2fc</th>\n",
       "      <th>anterior_mean</th>\n",
       "      <th>anterior_uncertainty</th>\n",
       "      <th>middle_mean</th>\n",
       "      <th>middle_uncertainty</th>\n",
       "      <th>scrambled_mean</th>\n",
       "      <th>scrambled_uncertainty</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>Adh</td>\n",
       "      <td>1.606079e-23</td>\n",
       "      <td>1.636547</td>\n",
       "      <td>0.000104</td>\n",
       "      <td>2.250335e-06</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>1.436296e-06</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>Amun</td>\n",
       "      <td>3.333921e-07</td>\n",
       "      <td>1.399808</td>\n",
       "      <td>0.000047</td>\n",
       "      <td>1.936108e-06</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>1.245992e-06</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>B52</td>\n",
       "      <td>5.202520e-20</td>\n",
       "      <td>1.012013</td>\n",
       "      <td>0.000129</td>\n",
       "      <td>2.444933e-06</td>\n",
       "      <td>0.000064</td>\n",
       "      <td>2.490844e-06</td>\n",
       "      <td>0.000078</td>\n",
       "      <td>0.000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>Bace</td>\n",
       "      <td>3.310686e-25</td>\n",
       "      <td>2.262314</td>\n",
       "      <td>0.000103</td>\n",
       "      <td>1.920798e-06</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>1.207054e-06</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>Bap55</td>\n",
       "      <td>4.707239e-06</td>\n",
       "      <td>1.424649</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>5.932932e-07</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>7.158474e-07</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6449</th>\n",
       "      <td>vkg</td>\n",
       "      <td>2.098643e-10</td>\n",
       "      <td>1.100035</td>\n",
       "      <td>0.000061</td>\n",
       "      <td>2.100971e-06</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>1.209043e-06</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6482</th>\n",
       "      <td>x16</td>\n",
       "      <td>4.566345e-11</td>\n",
       "      <td>1.321541</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>1.845769e-06</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>2.343774e-06</td>\n",
       "      <td>0.000031</td>\n",
       "      <td>0.000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6493</th>\n",
       "      <td>yps</td>\n",
       "      <td>1.593800e-25</td>\n",
       "      <td>1.124519</td>\n",
       "      <td>0.000148</td>\n",
       "      <td>3.077616e-06</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>1.905147e-06</td>\n",
       "      <td>0.000079</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6495</th>\n",
       "      <td>ytr</td>\n",
       "      <td>9.760874e-08</td>\n",
       "      <td>1.197079</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>1.226140e-06</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>1.645635e-06</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6501</th>\n",
       "      <td>zf30C</td>\n",
       "      <td>1.008911e-06</td>\n",
       "      <td>1.403969</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>9.660085e-07</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>8.177305e-07</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>128 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene         p_val    log2fc  anterior_mean  anterior_uncertainty  \\\n",
       "99      Adh  1.606079e-23  1.636547       0.000104          2.250335e-06   \n",
       "141    Amun  3.333921e-07  1.399808       0.000047          1.936108e-06   \n",
       "230     B52  5.202520e-20  1.012013       0.000129          2.444933e-06   \n",
       "240    Bace  3.310686e-25  2.262314       0.000103          1.920798e-06   \n",
       "244   Bap55  4.707239e-06  1.424649       0.000036          5.932932e-07   \n",
       "...     ...           ...       ...            ...                   ...   \n",
       "6449    vkg  2.098643e-10  1.100035       0.000061          2.100971e-06   \n",
       "6482    x16  4.566345e-11  1.321541       0.000060          1.845769e-06   \n",
       "6493    yps  1.593800e-25  1.124519       0.000148          3.077616e-06   \n",
       "6495    ytr  9.760874e-08  1.197079       0.000049          1.226140e-06   \n",
       "6501  zf30C  1.008911e-06  1.403969       0.000035          9.660085e-07   \n",
       "\n",
       "      middle_mean  middle_uncertainty  scrambled_mean  scrambled_uncertainty  \n",
       "99       0.000034        1.436296e-06        0.000065               0.000002  \n",
       "141      0.000018        1.245992e-06        0.000026               0.000002  \n",
       "230      0.000064        2.490844e-06        0.000078               0.000002  \n",
       "240      0.000022        1.207054e-06        0.000059               0.000001  \n",
       "244      0.000013        7.158474e-07        0.000020               0.000001  \n",
       "...           ...                 ...             ...                    ...  \n",
       "6449     0.000029        1.209043e-06        0.000036               0.000002  \n",
       "6482     0.000024        2.343774e-06        0.000031               0.000002  \n",
       "6493     0.000068        1.905147e-06        0.000079               0.000001  \n",
       "6495     0.000021        1.645635e-06        0.000022               0.000001  \n",
       "6501     0.000013        8.177305e-07        0.000021               0.000001  \n",
       "\n",
       "[128 rows x 9 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "anterior_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "129b0f33",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Adh,\n",
      "Amun,\n",
      "B52,\n",
      "Bace,\n",
      "Bap55,\n",
      "Bin1,\n",
      "Brd,\n",
      "CG11300,\n",
      "CG11370,\n",
      "CG11459,\n",
      "CG12115,\n",
      "CG12310,\n",
      "CG13044,\n",
      "CG13460,\n",
      "CG13461,\n",
      "CG13947,\n",
      "CG14265,\n",
      "CG14302,\n",
      "CG14332,\n",
      "CG14852,\n",
      "CG15093,\n",
      "CG15201,\n",
      "CG17278,\n",
      "CG17362,\n",
      "CG2310,\n",
      "CG30122,\n",
      "CG31460,\n",
      "CG31789,\n",
      "CG32198,\n",
      "CG33333,\n",
      "CG42748,\n",
      "CG42823,\n",
      "CG42834,\n",
      "CG43117,\n",
      "CG5789,\n",
      "CG6870,\n",
      "CG7290,\n",
      "CG7637,\n",
      "CG7953,\n",
      "CG8087,\n",
      "CG8289,\n",
      "CG8661,\n",
      "CG8664,\n",
      "CG8929,\n",
      "CG8997,\n",
      "CG9135,\n",
      "CG9672,\n",
      "CG9682,\n",
      "CG9686,\n",
      "Chrac-16,\n",
      "Cirl,\n",
      "Cont,\n",
      "CtBP,\n",
      "D1,\n",
      "Df31,\n",
      "DptA,\n",
      "EndoG,\n",
      "Ent2,\n",
      "Fas1,\n",
      "Fas3,\n",
      "Fbp1,\n",
      "GstD1,\n",
      "HP4,\n",
      "Hel25E,\n",
      "His2Av,\n",
      "His4r,\n",
      "HmgD,\n",
      "Hsp67Ba,\n",
      "ImpL2,\n",
      "Lam,\n",
      "LysS,\n",
      "Nxt1,\n",
      "O-fut1,\n",
      "Obp44a,\n",
      "Obp56d,\n",
      "Pebp1,\n",
      "Phae1,\n",
      "Phae2,\n",
      "Pig1,\n",
      "Prp8,\n",
      "RpII140,\n",
      "Set,\n",
      "Sin3A,\n",
      "SmD2,\n",
      "SmF,\n",
      "Srp54,\n",
      "Tom,\n",
      "Top2,\n",
      "TotA,\n",
      "Tsp42Ee,\n",
      "Uba2,\n",
      "baf,\n",
      "bnb,\n",
      "bowl,\n",
      "brat,\n",
      "cg,\n",
      "crol,\n",
      "deltaTry,\n",
      "drl,\n",
      "e(y)2,\n",
      "ed,\n",
      "emc,\n",
      "fax,\n",
      "hdc,\n",
      "jigr1,\n",
      "kuk,\n",
      "kune,\n",
      "l(3)neo38,\n",
      "miple2,\n",
      "mod,\n",
      "ng3,\n",
      "nocte,\n",
      "ogre,\n",
      "pck,\n",
      "pnut,\n",
      "pot,\n",
      "pros,\n",
      "psq,\n",
      "pzg,\n",
      "regucalcin,\n",
      "serp,\n",
      "smt3,\n",
      "stai,\n",
      "vkg,\n",
      "x16,\n",
      "yps,\n",
      "ytr,\n",
      "zf30C,\n"
     ]
    }
   ],
   "source": [
    "for g in anterior_df.gene.to_list():\n",
    "    print(g + ',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "d2be351f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>p_val</th>\n",
       "      <th>log2fc</th>\n",
       "      <th>anterior_mean</th>\n",
       "      <th>anterior_uncertainty</th>\n",
       "      <th>middle_mean</th>\n",
       "      <th>middle_uncertainty</th>\n",
       "      <th>scrambled_mean</th>\n",
       "      <th>scrambled_uncertainty</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>7SLRNA:CR32864</td>\n",
       "      <td>1.302461e-18</td>\n",
       "      <td>-2.362382</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>6.589371e-07</td>\n",
       "      <td>0.000133</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000081</td>\n",
       "      <td>0.000004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>7SLRNA:CR42652</td>\n",
       "      <td>1.425579e-13</td>\n",
       "      <td>-2.297837</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>1.508762e-06</td>\n",
       "      <td>0.000119</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>0.000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>573</th>\n",
       "      <td>CG12229</td>\n",
       "      <td>6.539440e-07</td>\n",
       "      <td>-1.776069</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>6.789324e-07</td>\n",
       "      <td>0.000074</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000046</td>\n",
       "      <td>0.000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>606</th>\n",
       "      <td>CG12605</td>\n",
       "      <td>5.901773e-06</td>\n",
       "      <td>-1.799067</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>8.576129e-07</td>\n",
       "      <td>0.000077</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000047</td>\n",
       "      <td>0.000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>935</th>\n",
       "      <td>CG14995</td>\n",
       "      <td>1.284945e-06</td>\n",
       "      <td>-1.456249</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>2.083878e-06</td>\n",
       "      <td>0.000112</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000069</td>\n",
       "      <td>0.000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1088</th>\n",
       "      <td>CG16719</td>\n",
       "      <td>8.069228e-07</td>\n",
       "      <td>-1.907928</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>9.124538e-07</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000044</td>\n",
       "      <td>0.000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1276</th>\n",
       "      <td>CG18568</td>\n",
       "      <td>6.744778e-09</td>\n",
       "      <td>-1.473717</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>1.507184e-06</td>\n",
       "      <td>0.000099</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000062</td>\n",
       "      <td>0.000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1320</th>\n",
       "      <td>CG2127</td>\n",
       "      <td>1.124712e-15</td>\n",
       "      <td>-1.509781</td>\n",
       "      <td>0.000084</td>\n",
       "      <td>1.411644e-06</td>\n",
       "      <td>0.000239</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.000148</td>\n",
       "      <td>0.000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1446</th>\n",
       "      <td>CG31029</td>\n",
       "      <td>1.798690e-16</td>\n",
       "      <td>-2.015389</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>2.036882e-06</td>\n",
       "      <td>0.000164</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000090</td>\n",
       "      <td>0.000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1517</th>\n",
       "      <td>CG31538</td>\n",
       "      <td>4.990744e-19</td>\n",
       "      <td>-1.969059</td>\n",
       "      <td>0.000062</td>\n",
       "      <td>1.911879e-06</td>\n",
       "      <td>0.000243</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000136</td>\n",
       "      <td>0.000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1721</th>\n",
       "      <td>CG33293</td>\n",
       "      <td>2.434321e-15</td>\n",
       "      <td>-1.829797</td>\n",
       "      <td>0.000052</td>\n",
       "      <td>1.350606e-06</td>\n",
       "      <td>0.000186</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>0.000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1910</th>\n",
       "      <td>CG3788</td>\n",
       "      <td>2.913507e-07</td>\n",
       "      <td>-1.522426</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>1.226787e-06</td>\n",
       "      <td>0.000120</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2228</th>\n",
       "      <td>CG4546</td>\n",
       "      <td>8.016226e-13</td>\n",
       "      <td>-1.978264</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>5.832109e-07</td>\n",
       "      <td>0.000126</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000077</td>\n",
       "      <td>0.000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2315</th>\n",
       "      <td>CG5050</td>\n",
       "      <td>2.564990e-13</td>\n",
       "      <td>-1.832361</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>1.192265e-06</td>\n",
       "      <td>0.000118</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.000004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2348</th>\n",
       "      <td>CG5280</td>\n",
       "      <td>1.002601e-10</td>\n",
       "      <td>-1.832857</td>\n",
       "      <td>0.000031</td>\n",
       "      <td>1.052728e-06</td>\n",
       "      <td>0.000110</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000064</td>\n",
       "      <td>0.000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2373</th>\n",
       "      <td>CG5458</td>\n",
       "      <td>1.756620e-09</td>\n",
       "      <td>-1.739680</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>1.581258e-06</td>\n",
       "      <td>0.000118</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000071</td>\n",
       "      <td>0.000004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2595</th>\n",
       "      <td>CG7387</td>\n",
       "      <td>3.577319e-07</td>\n",
       "      <td>-1.987368</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>7.796427e-07</td>\n",
       "      <td>0.000079</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000047</td>\n",
       "      <td>0.000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2685</th>\n",
       "      <td>CG8043</td>\n",
       "      <td>5.378801e-11</td>\n",
       "      <td>-1.632922</td>\n",
       "      <td>0.000044</td>\n",
       "      <td>1.214433e-06</td>\n",
       "      <td>0.000135</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000086</td>\n",
       "      <td>0.000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2783</th>\n",
       "      <td>CG8813</td>\n",
       "      <td>1.465183e-10</td>\n",
       "      <td>-1.642051</td>\n",
       "      <td>0.000037</td>\n",
       "      <td>1.404035e-06</td>\n",
       "      <td>0.000116</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000069</td>\n",
       "      <td>0.000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3573</th>\n",
       "      <td>Hsc70-2</td>\n",
       "      <td>1.161541e-09</td>\n",
       "      <td>-1.671024</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>1.608699e-06</td>\n",
       "      <td>0.000130</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000080</td>\n",
       "      <td>0.000004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3886</th>\n",
       "      <td>Mst77F</td>\n",
       "      <td>1.106199e-20</td>\n",
       "      <td>-2.086420</td>\n",
       "      <td>0.000053</td>\n",
       "      <td>2.069063e-06</td>\n",
       "      <td>0.000223</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000131</td>\n",
       "      <td>0.000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4277</th>\n",
       "      <td>Pof</td>\n",
       "      <td>9.891185e-14</td>\n",
       "      <td>-1.551586</td>\n",
       "      <td>0.000088</td>\n",
       "      <td>1.804912e-06</td>\n",
       "      <td>0.000259</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0.000160</td>\n",
       "      <td>0.000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4424</th>\n",
       "      <td>Rcd7</td>\n",
       "      <td>3.019822e-13</td>\n",
       "      <td>-1.940004</td>\n",
       "      <td>0.000030</td>\n",
       "      <td>6.129871e-07</td>\n",
       "      <td>0.000116</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000071</td>\n",
       "      <td>0.000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4597</th>\n",
       "      <td>Rrp6</td>\n",
       "      <td>3.635208e-06</td>\n",
       "      <td>-1.564058</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>1.037169e-06</td>\n",
       "      <td>0.000078</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000047</td>\n",
       "      <td>0.000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4902</th>\n",
       "      <td>Tektin-C</td>\n",
       "      <td>4.118945e-07</td>\n",
       "      <td>-1.530246</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>1.212931e-06</td>\n",
       "      <td>0.000112</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000069</td>\n",
       "      <td>0.000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5483</th>\n",
       "      <td>dpr17</td>\n",
       "      <td>1.292531e-08</td>\n",
       "      <td>-1.804620</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>1.678569e-06</td>\n",
       "      <td>0.000119</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000069</td>\n",
       "      <td>0.000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5585</th>\n",
       "      <td>fig</td>\n",
       "      <td>2.538363e-06</td>\n",
       "      <td>-1.875438</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>9.119244e-07</td>\n",
       "      <td>0.000078</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5634</th>\n",
       "      <td>gish</td>\n",
       "      <td>2.363128e-08</td>\n",
       "      <td>-1.635868</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>1.815591e-06</td>\n",
       "      <td>0.000150</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000089</td>\n",
       "      <td>0.000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6390</th>\n",
       "      <td>toc</td>\n",
       "      <td>1.315230e-06</td>\n",
       "      <td>-1.642260</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>1.499089e-06</td>\n",
       "      <td>0.000128</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000076</td>\n",
       "      <td>0.000004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6417</th>\n",
       "      <td>twin</td>\n",
       "      <td>4.651921e-17</td>\n",
       "      <td>-1.938624</td>\n",
       "      <td>0.000062</td>\n",
       "      <td>1.643360e-06</td>\n",
       "      <td>0.000238</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0.000138</td>\n",
       "      <td>0.000004</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene         p_val    log2fc  anterior_mean  \\\n",
       "16    7SLRNA:CR32864  1.302461e-18 -2.362382       0.000026   \n",
       "17    7SLRNA:CR42652  1.425579e-13 -2.297837       0.000024   \n",
       "573          CG12229  6.539440e-07 -1.776069       0.000021   \n",
       "606          CG12605  5.901773e-06 -1.799067       0.000022   \n",
       "935          CG14995  1.284945e-06 -1.456249       0.000041   \n",
       "1088         CG16719  8.069228e-07 -1.907928       0.000020   \n",
       "1276         CG18568  6.744778e-09 -1.473717       0.000036   \n",
       "1320          CG2127  1.124712e-15 -1.509781       0.000084   \n",
       "1446         CG31029  1.798690e-16 -2.015389       0.000041   \n",
       "1517         CG31538  4.990744e-19 -1.969059       0.000062   \n",
       "1721         CG33293  2.434321e-15 -1.829797       0.000052   \n",
       "1910          CG3788  2.913507e-07 -1.522426       0.000042   \n",
       "2228          CG4546  8.016226e-13 -1.978264       0.000032   \n",
       "2315          CG5050  2.564990e-13 -1.832361       0.000033   \n",
       "2348          CG5280  1.002601e-10 -1.832857       0.000031   \n",
       "2373          CG5458  1.756620e-09 -1.739680       0.000035   \n",
       "2595          CG7387  3.577319e-07 -1.987368       0.000020   \n",
       "2685          CG8043  5.378801e-11 -1.632922       0.000044   \n",
       "2783          CG8813  1.465183e-10 -1.642051       0.000037   \n",
       "3573         Hsc70-2  1.161541e-09 -1.671024       0.000041   \n",
       "3886          Mst77F  1.106199e-20 -2.086420       0.000053   \n",
       "4277             Pof  9.891185e-14 -1.551586       0.000088   \n",
       "4424            Rcd7  3.019822e-13 -1.940004       0.000030   \n",
       "4597            Rrp6  3.635208e-06 -1.564058       0.000026   \n",
       "4902        Tektin-C  4.118945e-07 -1.530246       0.000039   \n",
       "5483           dpr17  1.292531e-08 -1.804620       0.000034   \n",
       "5585             fig  2.538363e-06 -1.875438       0.000021   \n",
       "5634            gish  2.363128e-08 -1.635868       0.000048   \n",
       "6390             toc  1.315230e-06 -1.642260       0.000041   \n",
       "6417            twin  4.651921e-17 -1.938624       0.000062   \n",
       "\n",
       "      anterior_uncertainty  middle_mean  middle_uncertainty  scrambled_mean  \\\n",
       "16            6.589371e-07     0.000133            0.000003        0.000081   \n",
       "17            1.508762e-06     0.000119            0.000004        0.000068   \n",
       "573           6.789324e-07     0.000074            0.000002        0.000046   \n",
       "606           8.576129e-07     0.000077            0.000003        0.000047   \n",
       "935           2.083878e-06     0.000112            0.000003        0.000069   \n",
       "1088          9.124538e-07     0.000073            0.000002        0.000044   \n",
       "1276          1.507184e-06     0.000099            0.000003        0.000062   \n",
       "1320          1.411644e-06     0.000239            0.000006        0.000148   \n",
       "1446          2.036882e-06     0.000164            0.000004        0.000090   \n",
       "1517          1.911879e-06     0.000243            0.000005        0.000136   \n",
       "1721          1.350606e-06     0.000186            0.000005        0.000109   \n",
       "1910          1.226787e-06     0.000120            0.000003        0.000073   \n",
       "2228          5.832109e-07     0.000126            0.000003        0.000077   \n",
       "2315          1.192265e-06     0.000118            0.000003        0.000072   \n",
       "2348          1.052728e-06     0.000110            0.000002        0.000064   \n",
       "2373          1.581258e-06     0.000118            0.000004        0.000071   \n",
       "2595          7.796427e-07     0.000079            0.000003        0.000047   \n",
       "2685          1.214433e-06     0.000135            0.000003        0.000086   \n",
       "2783          1.404035e-06     0.000116            0.000003        0.000069   \n",
       "3573          1.608699e-06     0.000130            0.000004        0.000080   \n",
       "3886          2.069063e-06     0.000223            0.000004        0.000131   \n",
       "4277          1.804912e-06     0.000259            0.000007        0.000160   \n",
       "4424          6.129871e-07     0.000116            0.000003        0.000071   \n",
       "4597          1.037169e-06     0.000078            0.000002        0.000047   \n",
       "4902          1.212931e-06     0.000112            0.000003        0.000069   \n",
       "5483          1.678569e-06     0.000119            0.000003        0.000069   \n",
       "5585          9.119244e-07     0.000078            0.000003        0.000043   \n",
       "5634          1.815591e-06     0.000150            0.000004        0.000089   \n",
       "6390          1.499089e-06     0.000128            0.000004        0.000076   \n",
       "6417          1.643360e-06     0.000238            0.000007        0.000138   \n",
       "\n",
       "      scrambled_uncertainty  \n",
       "16                 0.000004  \n",
       "17                 0.000003  \n",
       "573                0.000002  \n",
       "606                0.000002  \n",
       "935                0.000003  \n",
       "1088               0.000002  \n",
       "1276               0.000002  \n",
       "1320               0.000003  \n",
       "1446               0.000003  \n",
       "1517               0.000005  \n",
       "1721               0.000003  \n",
       "1910               0.000003  \n",
       "2228               0.000002  \n",
       "2315               0.000004  \n",
       "2348               0.000002  \n",
       "2373               0.000004  \n",
       "2595               0.000003  \n",
       "2685               0.000002  \n",
       "2783               0.000003  \n",
       "3573               0.000004  \n",
       "3886               0.000003  \n",
       "4277               0.000003  \n",
       "4424               0.000003  \n",
       "4597               0.000002  \n",
       "4902               0.000002  \n",
       "5483               0.000002  \n",
       "5585               0.000002  \n",
       "5634               0.000003  \n",
       "6390               0.000004  \n",
       "6417               0.000004  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "middle_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "479a71e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "30"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(middle_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "ea428065",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "12588"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.where([g == 'jigr1' for g in gene_list])[0][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "326b1fb4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<matplotlib.lines.Line2D at 0x7f431a108520>]"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"plot example genes\"\"\"\n",
    "this_id = np.where([gg == 'pros' for gg in gene_list])[0][0]\n",
    "#this_id = 12588\n",
    "#expression = X[sel, this_id]\n",
    "expression = norm_reads[sel, this_id]\n",
    "\n",
    "plt.figure()\n",
    "plt.plot(ys, expression, 'ko', alpha=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "135d0de6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3195445/625980817.py:89: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/core/fromnumeric.py:3504: RuntimeWarning: Mean of empty slice.\n",
      "  return _methods._mean(a, axis=axis, dtype=dtype,\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/core/_methods.py:129: RuntimeWarning: invalid value encountered in divide\n",
      "  ret = ret.dtype.type(ret / rcount)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_3195445/625980817.py:119: RuntimeWarning: Mean of empty slice\n",
      "  return np.nanmean(line_dist_arr, axis=0), np.nanstd(line_dist_arr, axis=0)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Text(0, 0.5, 'binned expression')"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"plot example binned dist\"\"\"\n",
    "\n",
    "binned_exp, std_exp = bin_exp(expression, binned_ys, y_bins)\n",
    "#mean_scramble, std_scramble = get_line_dist_uncertainty(this_id, X_fb, binned_ys, y_bins, gene_list, filtered_gene_list, n_bootstraps=10)\n",
    "mean_scramble, std_scramble = get_line_dist_uncertainty(this_id, norm_reads[sel], binned_ys, y_bins, gene_list, filtered_gene_list, n_bootstraps=10)\n",
    "\n",
    "m = mean_scramble\n",
    "s = std_scramble\n",
    "l = m - s\n",
    "u = m + s\n",
    "\n",
    "plt.figure()\n",
    "plt.fill_between(y_bins, l, u, color='k', alpha=0.3)\n",
    "plt.errorbar(y_bins, binned_exp, std_exp, linewidth=4, color='c')\n",
    "#plt.errorbar(y_bins, binned_dptA, bootstrapped_uncertainty)\n",
    "\n",
    "plt.xlabel('ap')\n",
    "plt.ylabel('binned expression')\n",
    "#plt.xlim([-113, np.max(ys)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "840dfd13",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>long_name</th>\n",
       "      <th>protein_family</th>\n",
       "      <th>protein_class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DROME|FlyBase=FBgn0004362|UniProtKB=Q05783</td>\n",
       "      <td>HmgD</td>\n",
       "      <td>High mobility group protein D;HmgD;PTN00071477...</td>\n",
       "      <td>TRANSCRIPTION FACTOR A, MITOCHONDRIAL (PTHR481...</td>\n",
       "      <td>HMG box transcription factor(PC00024)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DROME|FlyBase=FBgn0035829|UniProtKB=Q8IQ92</td>\n",
       "      <td>HP4</td>\n",
       "      <td>FI01426p;HP4;;orthologs</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DROME|FlyBase=FBgn0014163|UniProtKB=Q95RI5</td>\n",
       "      <td>fax</td>\n",
       "      <td>Failed axon connections;fax;PTN000265169;ortho...</td>\n",
       "      <td>FAILED AXON CONNECTIONS (PTHR12289:SF78)</td>\n",
       "      <td>transporter(PC00227)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DROME|FlyBase=FBgn0004587|UniProtKB=P26686</td>\n",
       "      <td>B52</td>\n",
       "      <td>Serine-arginine protein 55;B52;PTN002751804;or...</td>\n",
       "      <td>SERINE-ARGININE PROTEIN 55 (PTHR23003:SF51)</td>\n",
       "      <td>RNA splicing factor(PC00148)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DROME|FlyBase=FBgn0029506|UniProtKB=Q7KJ73</td>\n",
       "      <td>Tsp42Ee</td>\n",
       "      <td>Tetraspanin;Tsp42Ee;PTN000445347;orthologs</td>\n",
       "      <td>IP01817P-RELATED (PTHR19282:SF521)</td>\n",
       "      <td>scaffold/adaptor protein(PC00226)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>DROME|FlyBase=FBgn0033313|UniProtKB=A1Z7G7</td>\n",
       "      <td>Cirl</td>\n",
       "      <td>Latrophilin Cirl;Cirl;PTN000244766;orthologs</td>\n",
       "      <td>LATROPHILIN CIRL (PTHR12011:SF347)</td>\n",
       "      <td>G-protein coupled receptor(PC00021)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>DROME|FlyBase=FBgn0000055|UniProtKB=P00334</td>\n",
       "      <td>Adh</td>\n",
       "      <td>Alcohol dehydrogenase;Adh;PTN000672297;orthologs</td>\n",
       "      <td>ALCOHOL DEHYDROGENASE (PTHR42901:SF1)</td>\n",
       "      <td>dehydrogenase(PC00092)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>DROME|FlyBase=FBgn0034390|UniProtKB=Q9V8M5</td>\n",
       "      <td>CG15093</td>\n",
       "      <td>Probable 3-hydroxyisobutyrate dehydrogenase, m...</td>\n",
       "      <td>3-HYDROXYISOBUTYRATE DEHYDROGENASE, MITOCHONDR...</td>\n",
       "      <td>dehydrogenase(PC00092)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>DROME|FlyBase=FBgn0263234|UniProtKB=Q9VKA9</td>\n",
       "      <td>Phae1</td>\n",
       "      <td>AT26814p;Phae1;PTN000668564;orthologs</td>\n",
       "      <td>AT26814P-RELATED (PTHR24276:SF91)</td>\n",
       "      <td>serine protease(PC00203)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>DROME|FlyBase=FBgn0032049|UniProtKB=Q9VLK3</td>\n",
       "      <td>Bace</td>\n",
       "      <td>Beta-site APP-cleaving enzyme, isoform A;Bace;...</td>\n",
       "      <td>BETA-SITE APP-CLEAVING ENZYME, ISOFORM A-RELAT...</td>\n",
       "      <td>aspartic protease(PC00053)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>89 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       gene_id     gene  \\\n",
       "0   DROME|FlyBase=FBgn0004362|UniProtKB=Q05783     HmgD   \n",
       "1   DROME|FlyBase=FBgn0035829|UniProtKB=Q8IQ92      HP4   \n",
       "2   DROME|FlyBase=FBgn0014163|UniProtKB=Q95RI5      fax   \n",
       "3   DROME|FlyBase=FBgn0004587|UniProtKB=P26686      B52   \n",
       "4   DROME|FlyBase=FBgn0029506|UniProtKB=Q7KJ73  Tsp42Ee   \n",
       "..                                         ...      ...   \n",
       "84  DROME|FlyBase=FBgn0033313|UniProtKB=A1Z7G7     Cirl   \n",
       "85  DROME|FlyBase=FBgn0000055|UniProtKB=P00334      Adh   \n",
       "86  DROME|FlyBase=FBgn0034390|UniProtKB=Q9V8M5  CG15093   \n",
       "87  DROME|FlyBase=FBgn0263234|UniProtKB=Q9VKA9    Phae1   \n",
       "88  DROME|FlyBase=FBgn0032049|UniProtKB=Q9VLK3     Bace   \n",
       "\n",
       "                                            long_name  \\\n",
       "0   High mobility group protein D;HmgD;PTN00071477...   \n",
       "1                             FI01426p;HP4;;orthologs   \n",
       "2   Failed axon connections;fax;PTN000265169;ortho...   \n",
       "3   Serine-arginine protein 55;B52;PTN002751804;or...   \n",
       "4          Tetraspanin;Tsp42Ee;PTN000445347;orthologs   \n",
       "..                                                ...   \n",
       "84       Latrophilin Cirl;Cirl;PTN000244766;orthologs   \n",
       "85   Alcohol dehydrogenase;Adh;PTN000672297;orthologs   \n",
       "86  Probable 3-hydroxyisobutyrate dehydrogenase, m...   \n",
       "87              AT26814p;Phae1;PTN000668564;orthologs   \n",
       "88  Beta-site APP-cleaving enzyme, isoform A;Bace;...   \n",
       "\n",
       "                                       protein_family  \\\n",
       "0   TRANSCRIPTION FACTOR A, MITOCHONDRIAL (PTHR481...   \n",
       "1                                                 NaN   \n",
       "2            FAILED AXON CONNECTIONS (PTHR12289:SF78)   \n",
       "3         SERINE-ARGININE PROTEIN 55 (PTHR23003:SF51)   \n",
       "4                  IP01817P-RELATED (PTHR19282:SF521)   \n",
       "..                                                ...   \n",
       "84                 LATROPHILIN CIRL (PTHR12011:SF347)   \n",
       "85              ALCOHOL DEHYDROGENASE (PTHR42901:SF1)   \n",
       "86  3-HYDROXYISOBUTYRATE DEHYDROGENASE, MITOCHONDR...   \n",
       "87                  AT26814P-RELATED (PTHR24276:SF91)   \n",
       "88  BETA-SITE APP-CLEAVING ENZYME, ISOFORM A-RELAT...   \n",
       "\n",
       "                            protein_class  \n",
       "0   HMG box transcription factor(PC00024)  \n",
       "1                                     NaN  \n",
       "2                    transporter(PC00227)  \n",
       "3            RNA splicing factor(PC00148)  \n",
       "4       scaffold/adaptor protein(PC00226)  \n",
       "..                                    ...  \n",
       "84    G-protein coupled receptor(PC00021)  \n",
       "85                 dehydrogenase(PC00092)  \n",
       "86                 dehydrogenase(PC00092)  \n",
       "87               serine protease(PC00203)  \n",
       "88             aspartic protease(PC00053)  \n",
       "\n",
       "[89 rows x 5 columns]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "panther_df = pd.read_csv('/home/brandon/Downloads/pantherGeneList(2).txt', sep=\"\\t\", header=None)\n",
    "panther_df = panther_df.rename({0: 'gene_id', 1: 'gene', 2: 'long_name', 3: 'protein_family', 4: 'protein_class'}, axis='columns')\n",
    "panther_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "b9faf0a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_TF(x):\n",
    "    return 'transcription' in x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "403e8e1d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'transcription' in classes[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "f765f5d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>long_name</th>\n",
       "      <th>protein_family</th>\n",
       "      <th>protein_class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DROME|FlyBase=FBgn0004362|UniProtKB=Q05783</td>\n",
       "      <td>HmgD</td>\n",
       "      <td>High mobility group protein D;HmgD;PTN00071477...</td>\n",
       "      <td>TRANSCRIPTION FACTOR A, MITOCHONDRIAL (PTHR481...</td>\n",
       "      <td>HMG box transcription factor(PC00024)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>DROME|FlyBase=FBgn0020309|UniProtKB=O61360</td>\n",
       "      <td>crol</td>\n",
       "      <td>CROL ALPHA;crol;PTN000694590;orthologs</td>\n",
       "      <td>ZINC FINGER PROTEIN 239-LIKE ISOFORM X1 (PTHR2...</td>\n",
       "      <td>C2H2 zinc finger transcription factor(PC00248)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>DROME|FlyBase=FBgn0259785|UniProtKB=Q9VP57</td>\n",
       "      <td>pzg</td>\n",
       "      <td>LD15904p;pzg;PTN001971830;orthologs</td>\n",
       "      <td>ENHANCER OF VARIEGATION 3-9-RELATED (PTHR24403...</td>\n",
       "      <td>C2H2 zinc finger transcription factor(PC00248)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>DROME|FlyBase=FBgn0020496|UniProtKB=O46036</td>\n",
       "      <td>CtBP</td>\n",
       "      <td>C-terminal-binding protein;CtBP;PTN000107759;o...</td>\n",
       "      <td>C-TERMINAL-BINDING PROTEIN (PTHR46029:SF7)</td>\n",
       "      <td>transcription cofactor(PC00217)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>DROME|FlyBase=FBgn0039350|UniProtKB=Q9VBP5</td>\n",
       "      <td>jigr1</td>\n",
       "      <td>Jing interacting gene regulatory 1, isoform A;...</td>\n",
       "      <td>JING INTERACTING GENE REGULATORY 1, ISOFORM A ...</td>\n",
       "      <td>DNA-binding transcription factor(PC00218)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>DROME|FlyBase=FBgn0000618|UniProtKB=Q9VYX1</td>\n",
       "      <td>e(y)2</td>\n",
       "      <td>Enhancer of yellow 2 transcription factor;e(y)...</td>\n",
       "      <td>TRANSCRIPTION AND MRNA EXPORT FACTOR ENY2 (PTH...</td>\n",
       "      <td>DNA-binding transcription factor(PC00218)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>DROME|FlyBase=FBgn0265276|UniProtKB=A0A0B4KGA3</td>\n",
       "      <td>l(3)neo38</td>\n",
       "      <td>Lethal (3) neo38, isoform L;l(3)neo38;PTN00222...</td>\n",
       "      <td>LETHAL (3) NEO38, ISOFORM L (PTHR23235:SF75)</td>\n",
       "      <td>C2H2 zinc finger transcription factor(PC00248)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>DROME|FlyBase=FBgn0263102|UniProtKB=A0A6M3Q7H0</td>\n",
       "      <td>psq</td>\n",
       "      <td>Pipsqueak, isoform O;psq;PTN002350197;orthologs</td>\n",
       "      <td>PIPSQUEAK, ISOFORM O (PTHR23110:SF109)</td>\n",
       "      <td>DNA-binding transcription factor(PC00218)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>DROME|FlyBase=FBgn0000289|UniProtKB=A8DYD1</td>\n",
       "      <td>cg</td>\n",
       "      <td>Combgap, isoform L;cg;PTN002811048;orthologs</td>\n",
       "      <td>COMBGAP, ISOFORM L-RELATED (PTHR24388:SF53)</td>\n",
       "      <td>C2H2 zinc finger transcription factor(PC00248)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>DROME|FlyBase=FBgn0000575|UniProtKB=P18491</td>\n",
       "      <td>emc</td>\n",
       "      <td>Protein extra-macrochaetae;emc;PTN000197694;or...</td>\n",
       "      <td>PROTEIN EXTRA-MACROCHAETAE (PTHR11723:SF17)</td>\n",
       "      <td>DNA-binding transcription factor(PC00218)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>DROME|FlyBase=FBgn0004595|UniProtKB=P29617</td>\n",
       "      <td>pros</td>\n",
       "      <td>Homeobox protein prospero;pros;PTN000258137;or...</td>\n",
       "      <td>HOMEOBOX PROTEIN PROSPERO (PTHR12198:SF0)</td>\n",
       "      <td>homeodomain transcription factor(PC00119)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>DROME|FlyBase=FBgn0270924|UniProtKB=Q9VL91</td>\n",
       "      <td>zf30C</td>\n",
       "      <td>LD23102p;zf30C;PTN002393436;orthologs</td>\n",
       "      <td>FI17813P1-RELATED (PTHR24409:SF440)</td>\n",
       "      <td>C2H2 zinc finger transcription factor(PC00248)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>DROME|FlyBase=FBgn0004893|UniProtKB=Q9VQU9</td>\n",
       "      <td>bowl</td>\n",
       "      <td>Protein bowel;bowl;PTN000367586;orthologs</td>\n",
       "      <td>PROTEIN BOWEL (PTHR14196:SF0)</td>\n",
       "      <td>zinc finger transcription factor(PC00244)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>DROME|FlyBase=FBgn0043001|UniProtKB=Q9V452</td>\n",
       "      <td>Chrac-16</td>\n",
       "      <td>Chromatin accessibility complex 16kD protein;C...</td>\n",
       "      <td>CHROMATIN ACCESSIBILITY COMPLEX PROTEIN 1 (PTH...</td>\n",
       "      <td>DNA-binding transcription factor(PC00218)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           gene_id       gene  \\\n",
       "0       DROME|FlyBase=FBgn0004362|UniProtKB=Q05783       HmgD   \n",
       "9       DROME|FlyBase=FBgn0020309|UniProtKB=O61360       crol   \n",
       "12      DROME|FlyBase=FBgn0259785|UniProtKB=Q9VP57        pzg   \n",
       "15      DROME|FlyBase=FBgn0020496|UniProtKB=O46036       CtBP   \n",
       "20      DROME|FlyBase=FBgn0039350|UniProtKB=Q9VBP5      jigr1   \n",
       "31      DROME|FlyBase=FBgn0000618|UniProtKB=Q9VYX1      e(y)2   \n",
       "35  DROME|FlyBase=FBgn0265276|UniProtKB=A0A0B4KGA3  l(3)neo38   \n",
       "40  DROME|FlyBase=FBgn0263102|UniProtKB=A0A6M3Q7H0        psq   \n",
       "41      DROME|FlyBase=FBgn0000289|UniProtKB=A8DYD1         cg   \n",
       "42      DROME|FlyBase=FBgn0000575|UniProtKB=P18491        emc   \n",
       "43      DROME|FlyBase=FBgn0004595|UniProtKB=P29617       pros   \n",
       "55      DROME|FlyBase=FBgn0270924|UniProtKB=Q9VL91      zf30C   \n",
       "64      DROME|FlyBase=FBgn0004893|UniProtKB=Q9VQU9       bowl   \n",
       "81      DROME|FlyBase=FBgn0043001|UniProtKB=Q9V452   Chrac-16   \n",
       "\n",
       "                                            long_name  \\\n",
       "0   High mobility group protein D;HmgD;PTN00071477...   \n",
       "9              CROL ALPHA;crol;PTN000694590;orthologs   \n",
       "12                LD15904p;pzg;PTN001971830;orthologs   \n",
       "15  C-terminal-binding protein;CtBP;PTN000107759;o...   \n",
       "20  Jing interacting gene regulatory 1, isoform A;...   \n",
       "31  Enhancer of yellow 2 transcription factor;e(y)...   \n",
       "35  Lethal (3) neo38, isoform L;l(3)neo38;PTN00222...   \n",
       "40    Pipsqueak, isoform O;psq;PTN002350197;orthologs   \n",
       "41       Combgap, isoform L;cg;PTN002811048;orthologs   \n",
       "42  Protein extra-macrochaetae;emc;PTN000197694;or...   \n",
       "43  Homeobox protein prospero;pros;PTN000258137;or...   \n",
       "55              LD23102p;zf30C;PTN002393436;orthologs   \n",
       "64          Protein bowel;bowl;PTN000367586;orthologs   \n",
       "81  Chromatin accessibility complex 16kD protein;C...   \n",
       "\n",
       "                                       protein_family  \\\n",
       "0   TRANSCRIPTION FACTOR A, MITOCHONDRIAL (PTHR481...   \n",
       "9   ZINC FINGER PROTEIN 239-LIKE ISOFORM X1 (PTHR2...   \n",
       "12  ENHANCER OF VARIEGATION 3-9-RELATED (PTHR24403...   \n",
       "15         C-TERMINAL-BINDING PROTEIN (PTHR46029:SF7)   \n",
       "20  JING INTERACTING GENE REGULATORY 1, ISOFORM A ...   \n",
       "31  TRANSCRIPTION AND MRNA EXPORT FACTOR ENY2 (PTH...   \n",
       "35       LETHAL (3) NEO38, ISOFORM L (PTHR23235:SF75)   \n",
       "40             PIPSQUEAK, ISOFORM O (PTHR23110:SF109)   \n",
       "41        COMBGAP, ISOFORM L-RELATED (PTHR24388:SF53)   \n",
       "42        PROTEIN EXTRA-MACROCHAETAE (PTHR11723:SF17)   \n",
       "43          HOMEOBOX PROTEIN PROSPERO (PTHR12198:SF0)   \n",
       "55                FI17813P1-RELATED (PTHR24409:SF440)   \n",
       "64                      PROTEIN BOWEL (PTHR14196:SF0)   \n",
       "81  CHROMATIN ACCESSIBILITY COMPLEX PROTEIN 1 (PTH...   \n",
       "\n",
       "                                     protein_class  \n",
       "0            HMG box transcription factor(PC00024)  \n",
       "9   C2H2 zinc finger transcription factor(PC00248)  \n",
       "12  C2H2 zinc finger transcription factor(PC00248)  \n",
       "15                 transcription cofactor(PC00217)  \n",
       "20       DNA-binding transcription factor(PC00218)  \n",
       "31       DNA-binding transcription factor(PC00218)  \n",
       "35  C2H2 zinc finger transcription factor(PC00248)  \n",
       "40       DNA-binding transcription factor(PC00218)  \n",
       "41  C2H2 zinc finger transcription factor(PC00248)  \n",
       "42       DNA-binding transcription factor(PC00218)  \n",
       "43       homeodomain transcription factor(PC00119)  \n",
       "55  C2H2 zinc finger transcription factor(PC00248)  \n",
       "64       zinc finger transcription factor(PC00244)  \n",
       "81       DNA-binding transcription factor(PC00218)  "
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "classes = panther_df.protein_class.to_list()\n",
    "classes = [str(c) for c in classes]\n",
    "TFs = ['transcription' in c for c in classes]\n",
    "\n",
    "tf_df = panther_df[TFs]\n",
    "tf_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "a94d392d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3195445/625980817.py:89: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/core/fromnumeric.py:3504: RuntimeWarning: Mean of empty slice.\n",
      "  return _methods._mean(a, axis=axis, dtype=dtype,\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/core/_methods.py:129: RuntimeWarning: invalid value encountered in divide\n",
      "  ret = ret.dtype.type(ret / rcount)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_3195445/625980817.py:89: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/core/fromnumeric.py:3504: RuntimeWarning: Mean of empty slice.\n",
      "  return _methods._mean(a, axis=axis, dtype=dtype,\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/core/_methods.py:129: RuntimeWarning: invalid value encountered in divide\n",
      "  ret = ret.dtype.type(ret / rcount)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_3195445/625980817.py:89: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/core/fromnumeric.py:3504: RuntimeWarning: Mean of empty slice.\n",
      "  return _methods._mean(a, axis=axis, dtype=dtype,\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/core/_methods.py:129: RuntimeWarning: invalid value encountered in divide\n",
      "  ret = ret.dtype.type(ret / rcount)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_3195445/625980817.py:89: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/core/fromnumeric.py:3504: RuntimeWarning: Mean of empty slice.\n",
      "  return _methods._mean(a, axis=axis, dtype=dtype,\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/core/_methods.py:129: RuntimeWarning: invalid value encountered in divide\n",
      "  ret = ret.dtype.type(ret / rcount)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_3195445/625980817.py:89: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/core/fromnumeric.py:3504: RuntimeWarning: Mean of empty slice.\n",
      "  return _methods._mean(a, axis=axis, dtype=dtype,\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/core/_methods.py:129: RuntimeWarning: invalid value encountered in divide\n",
      "  ret = ret.dtype.type(ret / rcount)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_3195445/625980817.py:89: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/core/fromnumeric.py:3504: RuntimeWarning: Mean of empty slice.\n",
      "  return _methods._mean(a, axis=axis, dtype=dtype,\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/core/_methods.py:129: RuntimeWarning: invalid value encountered in divide\n",
      "  ret = ret.dtype.type(ret / rcount)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_3195445/625980817.py:89: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/core/fromnumeric.py:3504: RuntimeWarning: Mean of empty slice.\n",
      "  return _methods._mean(a, axis=axis, dtype=dtype,\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/core/_methods.py:129: RuntimeWarning: invalid value encountered in divide\n",
      "  ret = ret.dtype.type(ret / rcount)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_3195445/625980817.py:89: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/core/fromnumeric.py:3504: RuntimeWarning: Mean of empty slice.\n",
      "  return _methods._mean(a, axis=axis, dtype=dtype,\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/core/_methods.py:129: RuntimeWarning: invalid value encountered in divide\n",
      "  ret = ret.dtype.type(ret / rcount)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_3195445/625980817.py:89: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/core/fromnumeric.py:3504: RuntimeWarning: Mean of empty slice.\n",
      "  return _methods._mean(a, axis=axis, dtype=dtype,\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/core/_methods.py:129: RuntimeWarning: invalid value encountered in divide\n",
      "  ret = ret.dtype.type(ret / rcount)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_3195445/625980817.py:89: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/core/fromnumeric.py:3504: RuntimeWarning: Mean of empty slice.\n",
      "  return _methods._mean(a, axis=axis, dtype=dtype,\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/core/_methods.py:129: RuntimeWarning: invalid value encountered in divide\n",
      "  ret = ret.dtype.type(ret / rcount)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_3195445/625980817.py:89: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/core/fromnumeric.py:3504: RuntimeWarning: Mean of empty slice.\n",
      "  return _methods._mean(a, axis=axis, dtype=dtype,\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/core/_methods.py:129: RuntimeWarning: invalid value encountered in divide\n",
      "  ret = ret.dtype.type(ret / rcount)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_3195445/625980817.py:89: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/core/fromnumeric.py:3504: RuntimeWarning: Mean of empty slice.\n",
      "  return _methods._mean(a, axis=axis, dtype=dtype,\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/core/_methods.py:129: RuntimeWarning: invalid value encountered in divide\n",
      "  ret = ret.dtype.type(ret / rcount)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3195445/625980817.py:89: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/core/fromnumeric.py:3504: RuntimeWarning: Mean of empty slice.\n",
      "  return _methods._mean(a, axis=axis, dtype=dtype,\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/core/_methods.py:129: RuntimeWarning: invalid value encountered in divide\n",
      "  ret = ret.dtype.type(ret / rcount)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/tmp/ipykernel_3195445/625980817.py:89: RuntimeWarning: Mean of empty slice\n",
      "  binned_exp[i] = np.nanmean(these_expression_levels)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/core/fromnumeric.py:3504: RuntimeWarning: Mean of empty slice.\n",
      "  return _methods._mean(a, axis=axis, dtype=dtype,\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/core/_methods.py:129: RuntimeWarning: invalid value encountered in divide\n",
      "  ret = ret.dtype.type(ret / rcount)\n",
      "/home/brandon/anaconda3/envs/ome_zarr/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1879: RuntimeWarning: Degrees of freedom <= 0 for slice.\n",
      "  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n"
     ]
    }
   ],
   "source": [
    "tf_genes = tf_df.gene.to_list()\n",
    "anterior_genes = anterior_df.gene.to_list()\n",
    "\n",
    "plt.figure()\n",
    "for i, g in enumerate(tf_genes):\n",
    "    this_id = np.where([gg == g for gg in gene_list])[0][0]\n",
    "    expression = norm_reads[sel, this_id]\n",
    "\n",
    "    binned_exp, _ = bin_exp(expression, binned_ys, y_bins)\n",
    "    \n",
    "    plt.plot(y_bins, binned_exp / np.nanmax(binned_exp), linewidth=2, alpha=0.8)\n",
    "\n",
    "plt.xlabel('ap')\n",
    "plt.ylabel('binned expression')\n",
    "plt.xlim([-130, 185])\n",
    "plt.yscale('linear')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "f4986798",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "184.0076"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ys.max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2217cfd0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (ome_zarr)",
   "language": "python",
   "name": "ome_zarr"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
